### PR TITLE
Merge ArrayMapNamingStrategy into MapNamingStrategy

### LINKED ISSUE #9: MapNamingStrategy vs. ArrayMapNamingStrategy
Do we need both?  From what I can deduce (from source and [the PR that introduced ArrayMapNamingStrategy](https://github.com/zendframework/zf2/pull/6197)) they're nearly identical except:
- MapNamingStrategy's argument is a hydration map and ArrayMapNamingStrategy is an extraction map.  Both then flip the array to get the mapping for the opposite direction. 
- MapNamingStrategy allows overriding the reverse map, in case you want an asymmetrical mapping.

### PR DESCRIPTION
**Pull request by weierophinney**
Per #9, the two perform almost identical functionality; the main difference is _which argument is first in the constructor_, and the fact that `MapNamingStrategy` allowed a second argument for the opposite direction.

This patch merges the two into `MapNamingStrategy`. The class is now final (as was `ArrayMapNamingStrategy`, and accepts either a `$hydrationMap` or an `$extractionMap` or both. If both are missing, it raises an exception. If one or the other is missing, the value becomes that of the `array_flip` of the other value present.

The patch includes a migration document section detailing the changes, and expands the `MapNamingStrategy` documentation to cover all use cases.

Fixes #9

### CODE REVIEW

File: src/NamingStrategy/MapNamingStrategy.php [hunk 1]
```diff
@@ -9,83 +9,100 @@
...
-        $this->reverse = $reverse ?: $this->flipMapping($mapping);
+        if (null === $hydrationMap && null === $extractionMap) {
+            throw new Exception\InvalidArgumentException(sprintf(
+                '%s requires one or both of an array $hydrationMap and array $extractionMap;'
+                . ' neither provided',
+                __CLASS__
+            ));
+        }
+
+        if (null === $extractionMap && is_array($hydrationMap)) {
```
Comment (Ocramius) [inline src/NamingStrategy/MapNamingStrategy.php]: `is_array` wasn't imported, but a not null check would also suffice here

File: src/NamingStrategy/MapNamingStrategy.php [hunk 2]
```diff
@@ -9,83 +9,100 @@
...
+            $extractionMap = $this->flipMapping($hydrationMap);
+        }
+
+        if (null === $hydrationMap && is_array($extractionMap)) {
```
Comment (Ocramius) [inline src/NamingStrategy/MapNamingStrategy.php]: Same as above

File: src/NamingStrategy/MapNamingStrategy.php [hunk 3]
```diff
@@ -9,83 +9,100 @@
...
+     *     translation of extracted field names. If not provided, the result of
+     *     an array_flip($hydrationMap) will be used.
+     * @throws Exception\InvalidArgumentException if neither $hydrationMap nor
+     *     $extractionMap are provided.
+     * @throws Exception\InvalidArgumentException when flipping either the
+     *     $hydrationMap or $extractionMap, and any value is a non-string,
+     *     non-int value.
      */
-    public function __construct(array $mapping, ?array $reverse = null)
+    public function __construct(?array $hydrationMap = null, ?array $extractionMap = null)
```
Comment (Ocramius) [inline src/NamingStrategy/MapNamingStrategy.php]: Instead of having a single ctor with an edge case, what about having 3 named ctors that don't have the edge case at all?

File: src/NamingStrategy/MapNamingStrategy.php [hunk 4]
```diff
@@ -9,83 +9,100 @@
...
-     *
      * @var string[]
      */
-    protected $reverse = [];
+    private $hydrationMap = [];
 
     /**
-     * @param string[]      $mapping Map for name conversion on hydration
-     * @param null|string[] $reverse Reverse map for name conversion on extraction
+     * @param null|string[] $hydrationMap A map of string keys and values for
```
Comment (Ocramius) [inline src/NamingStrategy/MapNamingStrategy.php]: In this case, the `null|array<string, string>` notation is quite important, as the type of key is relevant to us.

File: src/NamingStrategy/MapNamingStrategy.php [hunk 5]
```diff
@@ -9,83 +9,100 @@
...
+    public function extract(string $name, ?object $object = null) : string
     {
-        array_walk($array, function ($value) {
-            if (! is_string($value) && ! is_int($value)) {
-                throw new InvalidArgumentException('Mapping array can\'t be flipped because of invalid value');
-            }
-        });
-
-        return array_flip($array);
+        return array_key_exists($name, $this->extractionMap)
```
Comment (Ocramius) [inline src/NamingStrategy/MapNamingStrategy.php]: Since we know that all `$this->extractionMap` is `array<string, string>`, this method can be simplified to:

```php
return $this->extractionMap[$name] ?? $name;
```

File: src/NamingStrategy/MapNamingStrategy.php [hunk 6]
```diff
@@ -9,83 +9,100 @@
...
      * {@inheritDoc}
      */
     public function hydrate(string $name, ?array $data = null) : string
     {
-        if (array_key_exists($name, $this->mapping)) {
-            return $this->mapping[$name];
-        }
-
-        return $name;
+        return array_key_exists($name, $this->hydrationMap)
```
Comment (Ocramius) [inline src/NamingStrategy/MapNamingStrategy.php]: Same as above with simplifications due to type assumptions

File: src/NamingStrategy/MapNamingStrategy.php [hunk 7]
```diff
@@ -9,83 +9,100 @@
...
+     *     a non-string, non-int value.
      */
-    public function extract(string $name, ?object $object = null) : string
+    private function flipMapping(array $array) : array
     {
-        if (array_key_exists($name, $this->reverse)) {
-            return $this->reverse[$name];
-        }
+        array_walk($array, function ($value) {
+            if (! is_string($value) && ! is_int($value)) {
```
Comment (Ocramius) [inline src/NamingStrategy/MapNamingStrategy.php]: `is_int()` isn't acceptable according to mapping type definitions

File: test/NamingStrategy/MapNamingStrategyTest.php [hunk 8]
```diff
@@ -9,50 +9,93 @@
...
+        $this->expectException(Exception\InvalidArgumentException::class);
+        $this->expectExceptionMessage('requires one or both of an array $hydrationMap and array $extractionMap');
 
-        $this->assertEquals('bar', $namingStrategy->hydrate('foo'));
-        $this->assertEquals('foo', $namingStrategy->extract('bar'));
+        new MapNamingStrategy();
     }
 
-    public function testHydrateAndExtractMaps()
+    public function invalidMapValues() : iterable
```
Comment (Ocramius) [inline test/NamingStrategy/MapNamingStrategyTest.php]: You also want to test with string keys. Yes, integer-ish keys are auto-vast to integers by the horrors of the PHP engine, but it is an interesting edge case to cover.

File: src/NamingStrategy/MapNamingStrategy.php (hunk 1 above)
Comment (weierophinney) [reply src/NamingStrategy/MapNamingStrategy.php]: phpstan complained without the `is_array()` check; I actually added it in after-the-fact due to a warning phpstan raised.

File: src/NamingStrategy/MapNamingStrategy.php (hunk 3 above)
Comment (weierophinney) [reply src/NamingStrategy/MapNamingStrategy.php]: Good plan - I'll work on that today.

File: src/NamingStrategy/MapNamingStrategy.php (hunk 1 above)
Comment (weierophinney) [reply src/NamingStrategy/MapNamingStrategy.php]: Resolved by using named constructors.

**Comment by weierophinney** [comment]
@Ocramius I've created three named constructors:

- `createFromExtractionMap(array $extractionMap)`
- `createFromHydrationMap(array $hydrationMap)`
- `createFromAssymetricMap(array $extractionMap, array $hydrationMap`

This solved the various issues I had with phpstan, and, as you note, it makes the behavior far more clear.

I've added tests to check for non-string keys as well; these are _almost_ like those for values, but I had to remove the `array` and `object` cases, as PHP considers those invalid index values and throws fatals anyways.

Thanks for the review!

**Comment by Ocramius** [comment]
Moved discussion point into https://github.com/zendframework/zend-hydrator/issues/85

**Comment by Ocramius** [comment]
Feedback handled in #86
